hnn_core/gui/_viz_manager.py

```diff
@@ -116,17 +117,21 @@ def unlink_relink(attribute):
     def _unlink_relink(f):
         @wraps(f)
         def wrapper(self, *args, **kwargs):
-            # Unlink the widgets using the provided link object
-            link_attribute: link = getattr(self, attribute)
-            link_attribute.unlink()
+            try:
```

Comment:
why do you need try-except? it's not a good idiom in Python ... replacing with `if`-`else` is preferable when possible

Reply:
From Zen of Python:

> Errors should never pass silently.
> Unless explicitly silenced.

see also: https://peps.python.org/pep-0008/#programming-recommendations

Reply:
There is internal code in the `unlink` call doesnt check the state of some attributes, causing  `out of bound`s exceptions that are unhandled by hnn or the traitlets package. Additionally, the `f` handler call also throws exceptions unhandled by hnn. I can explicitly refer this exceptions in the `try-except` block or use `if-else` , but what if more exceptions come out? I  am open to suggestions. Also, I can jut remove this block and create an issue for a future PR.

Reply:
Camilo and I discussed this. It is useful as we are actively developing new features for the GUI over the next few months. Often times we're not debugging through a live GUI.  We're going to keep it commented in pushes but will remove it entirely once the GUI is in production.

Reply:
If you need it for debugging, you could use a `debug` flag. I would rather have it uncommented than not. But with an explanation why it's needed in the comments. I also realized that we're missing explanation for the unlink/relink logic in the code itself. I had to dig through github issues to understand why that exists. We should have a two-line comment with a pointer to an open github issue (on ipywidgets) suggesting when this logic can be removed. I worry we're adding patch on top of patch.

Reply:
https://www.redhat.com/en/blog/why-upstream-contributions-matter-when-developing-open-source-nfv-solutions 

even if we don't contribute to `ipywidgets` it's important to highlight an issue/bug so we know it will get fixed in the future ...

Reply:
I don't think the unlink-relink solution is due to an ipywidget bug. Validation logic was refined for ipywidget 8.0 which caused this issue in the GUI to crop up. We use linking between tab groups so that when a figure tab is selected the figure-control tab of that figure is automatically selected. 

If you were to add a new figure tab, you'll have a brief moment where the figure tabs will have a single tab more than the figure-control tab.  If the tab groups are still linked this will cause an index error when the figure-controls tries to select a tab it doesn't currently have. Therefore you must unlink the two tab groups, add the new tabs to both groups, then relink the groups. 

It is currently invoked here in the code:
https://github.com/jonescompneurolab/hnn-core/blob/3be7a2d394d4ad181d512b57fb3cd8df46b4a035/hnn_core/gui/_viz_manager.py#L784-L792

Would adding this comment there be sufficient? 

```python
# Unlinks figure and figure-controls tab groups so new tabs can be added without index error.
```

Reply:
okay understood. But I don't understand the following:

> There is internal code in the unlink call doesnt check the state of some attributes, causing out of bounds exceptions that are unhandled by hnn or the traitlets package. 

It seems to suggest a bug in ipywidgets? Can I see the traceback?

Also, looking at the code above, why do you have decorator when it's used in a single function? Do we envision using this logic elsewhere? I would reduce complexity if not and get rid of decorator altogether ... it raises barrier to entry for new contributors

Reply:
The first bug is located in the link_attribute.unlink() method, which calls the `traitlets `package, a dependency of `ipywidgets`. This is the workflow:

- Make an empty figure with no sim or loaded data in the app.
- Load or simulate data.

This is the traceback:
```
'Traceback (most recent call last):
 File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 126, in wrapper
 link_attribute.unlink()
 File "/mnt/c/Projects/Github/hnn-core-ccv/venv310/lib/python3.10/site-packages/traitlets/traitlets.py", line 344, in unlink
 self.source[0].unobserve(self._update_target, names=self.source[1])
 File "/mnt/c/Projects/Github/hnn-core-ccv/venv310/lib/python3.10/site-packages/traitlets/traitlets.py", line 1698, in unobserve
 self._remove_notifiers(handler, name, type)
 File "/mnt/c/Projects/Github/hnn-core-ccv/venv310/lib/python3.10/site-packages/traitlets/traitlets.py", line 1592, in _remove_notifiers
 self._trait_notifiers[name][type].remove(handler)\nValueError: list.remove(x): x not in list'
```
The second bug is when you go to the visualization tab with no loaded data, and click the button 'Make figure'. This is the traceback:
```

 'Traceback (most recent call last):  
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 129, in wrapper    
  result = f(self, *args, **kwargs)
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 811, in add_figure
  _add_figure(None,
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 685, in _add_figure
  _add_axes_controls(widgets, data, fig=fig, axd=axd)
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 627, in _add_axes_controls
  children = [
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 628, in <listcomp>
  _get_ax_control(widgets, data, fig_idx=fig_idx, fig=fig, ax=ax)
  File "/mnt/c/Projects/Github/hnn-core-ccv/hnn_core/gui/_viz_manager.py", line 432, in _get_ax_control
  sim_name_default = simulation_names[-1]\nIndexError: tuple index out of range'
```

Reply:
hard for me to suggest what's the fix without trying myself (I'm AFK for a few days) but the code doesn't smell right ... it feels like the underlying issue requires reworking the logic of the setting up the data structures to manipulate the visualization tab. I would suggest ditching the decorator and rather handling this corner case correctly.

Reply:
Seems like this is an edge case that isn't relevant to this PR. @kmilo9999 Would you add an issue about this edge case and link to this discussion?  For now I think we should merge this PR as it is blocking subsequent development.

Reply:
#742

Reply:
The try-except makes me nervous. It will suppress genuine errors. Please remove it and address what's necessary in a separate PR ... one never finds the time to go back and fix things.

Reply:
I'm confused. This is actually going to print all errors and exceptions and not suppress any errors. 

```python
try:
    function()
except Exception as e:
    print(f"{e}")
```

You could add a `raise e` if you want the program to stop upon finding an error.  Though that's not ideal in the GUI as you will need to restart the GUI. 

In production code you typically don't want to use `except Exception`, but something more specific like `except ValueError`, but that's only once you know the common exceptions that occur. The general `Exception` is good during development to find out what those common errors are.

Reply:
Anyway, this logic is pretty minor and quick to add back while debugging so we should just remove it and not discuss it any further. But I did create issue #737 this week to remind CCV to review our work and remove any dev-related helper clauses/function before release that may slip through PRs.

Reply:
Thanks :) The print statement isn't the same level as an error but it's certainly better than nothing. I would have preferred to have the errors thrown using the ipywidgets logging infrastructure as everything would then be visible in the same consolidated output panel. We can discuss more once I'm back from vacation. In the meantime, if you can inject the try-except as you need it, that would be great!